### PR TITLE
Use b3dm tilesets for classification

### PR DESCRIPTION
**Pull request by bagnell**
This adds a `classificationType` option to `Cesium3DTileset` that will turn the glTF of a b3dm into a classification volume.

`ClassificationModel` has support for some of the same options that `Model` does (e.g. terrain clamping). Can this be removed or should we make this part of the public API?

Limitations:
* only works for b3dm
* all indices with the same batch id must be contiguous in memory
* ignores all shaders and techniques
* the only extension supported is `WEB3D_quantized_attributes`

Note that this is a PR into the vector-tiles branch.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Signed CLA is on file.

@bagnell, thanks for the pull request! Maintainers, we have a signed CLA from @bagnell, so you can review this at any time.

:warning: I noticed that [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md) has not been updated. If this change updates the public API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and comment on this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by pjcozzi** [comment]
Limitations all sound OK to me.  Later I could see users wanting to classify with i3dm - it is is kind of in between `b3dm` and `geom`.

**Comment by pjcozzi** [comment]
> ClassificationModel has support for some of the same options that Model does (e.g. terrain clamping). Can this be removed or should we make this part of the public API?

I don't think `ClassificationModel` needs to be exposed through the public API; Cesium API users will use the `ClassificationPrimitive` if they are not using 3D Tiles...I think.

So it only needs to have the API needed to support `Cesium3DTileset`, which does not have terrain clamping I believe.

File: Source/Scene/Cesium3DTileset.js [hunk 1]
```diff
@@ -1053,6 +1046,19 @@ define([
             get : function() {
                 return this._statistics;
             }
+        },
+
+        /**
+         * Determines whether terrain, 3D Tiles or both will be classified by this tileset.
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Label as `@readonly`.

File: Source/Scene/Cesium3DTileset.js [hunk 2]
```diff
@@ -222,6 +222,8 @@ define([
 
         this._readyPromise = when.defer();
 
+        this._classificationType = options.classificationType;
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Is this in the reference doc?

**Comment by pjcozzi** [comment]
Update CHANGES.md.

**Comment by pjcozzi** [comment]
Add a Sandcastle example.

File: Source/Scene/Vector3DTilePrimitive.js [hunk 3]
```diff
@@ -153,8 +153,18 @@ define([
          * @type {ClassificationType}
          * @default ClassificationType.CESIUM_3D_TILE
          */
-        this.classificationType = ClassificationType.CESIUM_3D_TILE;
-        this._classificationType = this.classificationType;
+        this.classificationType = defaultValue(options.classificationType, ClassificationType.CESIUM_3D_TILE);
```
Comment (pjcozzi) [inline Source/Scene/Vector3DTilePrimitive.js]: Add to reference doc even though the class is private.

File: Source/Scene/Cesium3DTileset.js (hunk 2 above)
Comment (pjcozzi) [reply Source/Scene/Cesium3DTileset.js]: Also include the limitations in the reference doc.  It might be easiest to add this once in the property getter and then reference it from the constructor function.

**Comment by pjcozzi** [comment]
Add unit tests.

File: Source/Scene/Batched3DModel3DTileContent.js [hunk 4]
```diff
@@ -354,29 +377,50 @@ define([
...
+                classificationShaderLoaded : getClassificationFragmentShaderCallback(content),
+                uniformMapLoaded : batchTable.getUniformMapCallback(),
+                pickVertexShaderLoaded : getPickVertexShaderCallback(content),
+                pickFragmentShaderLoaded : batchTable.getPickFragmentShaderCallback(),
+                pickUniformMapLoaded : batchTable.getPickUniformMapCallback(),
+                addBatchIdToGeneratedShaders : (batchLength > 0), // If the batch table has values in it, generated shaders will need a batchId attribute
+                pickObject : pickObject
+            });
+        } else {
+            content._model = new ClassificationModel({
```
Comment (pjcozzi) [inline Source/Scene/Batched3DModel3DTileContent.js]: Can you add a code comment about how this is a transcode - with limitations on the input glTF - so that we can use the `geom` rebatching rendering pipeline?

File: Source/Scene/ClassificationModel.js [hunk 5]
```diff
@@ -0,0 +1,2482 @@
+define([
```
Comment (pjcozzi) [inline Source/Scene/ClassificationModel.js]: Looks like it does more than it needs to given the limitations classification has.  Are you sure there aren't other limitations we can add to make this even more simple, e.g., vertex formats, removing the cache (is it needed), etc.

Then perhaps factor out some common functions to share with `Model`?

**Comment by bagnell** [comment]
@pjcozzi This is ready for another look.

**Comment by pjcozzi** [comment]
Do you get test failures like this:

image

when running http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/b3dm-class/Specs/SpecRunner.html

I get a total of

> 9590 specs, 96 failures, 6 pending specs

Some may just be before I am running from S3.

MacBook Pro / AMD.

File: Apps/Sandcastle/gallery/3D Tiles Photogrammetry Classification.html [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    viewer.camera.viewBoundingSphere(boundingSphere, new Cesium.HeadingPitchRange(0.0, -0.5, boundingSphere.radius));
+    viewer.camera.lookAtTransform(Cesium.Matrix4.IDENTITY);
+}).otherwise(function(error) {
+    throw(error);
+});
+
+var classification = viewer.scene.primitives.add(new Cesium.Cesium3DTileset({
+    url: 'https://beta.cesium.com/api/assets/3486?access_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJqdGkiOiIzZjM4MjljZi0xNTAzLTQ0MzctODhlZi0zOTU3Njg5ODA1Y2QiLCJpZCI6OCwiaWF0IjoxNDgxODI5ODMyfQ.dYlV8_PoXcFNlPHGook5kMzuJMS-Bb9DCMzI1mFVvgE',
+    classificationType : Cesium.ClassificationType.CESIUM_3D_TILE,
+    skipLevelOfDetail : false
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/3D Tiles Photogrammetry Classification.html]: Can this be omitted?

I thought the plan was to update the documentation for `skipLevelOfDetail` so that it was a "hint" - and then to just implicitly disable it when a tileset is used for classification.

File: Apps/Sandcastle/gallery/3D Tiles Photogrammetry Classification.html [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+tileset.readyPromise.then(function() {
+    var boundingSphere = tileset.boundingSphere;
+    viewer.camera.viewBoundingSphere(boundingSphere, new Cesium.HeadingPitchRange(0.0, -0.5, boundingSphere.radius));
+    viewer.camera.lookAtTransform(Cesium.Matrix4.IDENTITY);
+}).otherwise(function(error) {
+    throw(error);
+});
+
+var classification = viewer.scene.primitives.add(new Cesium.Cesium3DTileset({
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/3D Tiles Photogrammetry Classification.html]: What does this tileset look like without classification.  Is it worth adding a comment to say to comment out the `classificationType` line to see the geometry used to classify?

The other option is to create two tilesets, one with and without the classification, and then change their `show` via a checkbox, but that could make the example too complex.

File: CHANGES.md [hunk 8]
```diff
@@ -9,6 +9,13 @@ Change Log
...
     * Added `PlaneGeometry`, `PlaneOutlineGeometry`, `PlaneGeometryUpdater`, and `PlaneOutlineGeometryUpdater` classes to render plane primitives. [#5996](https://github.com/AnalyticalGraphicsInc/cesium/pull/5996)
     * Added `PlaneGraphics` class and `plane` property to `Entity`. [#5996](https://github.com/AnalyticalGraphicsInc/cesium/pull/5996)
+* Added `AttributeCompression.zigZagDeltaDecode` which will decode delta and ZigZag encoded buffers in place.
+* Added `pack` and `unpack` functions to `OrientedBoundingBox` for packing to and unpacking from a flat buffer.
+* Added experimental support for [3D Tiles Vector Data](https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/vector-tiles/TileFormats/VectorData) ([#4665](https://github.com/AnalyticalGraphicsInc/cesium/pull/4665)). The new and modified Cesium APIs are:
+   * `Cesium3DTileStyle` has expanded for styling point features. See the [styling specification](https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/vector-tiles/Styling#vector-data) for details.
+   * `Cesium3DTileFeature` can modify `color` and `show` properties for polygon, polyline, geometry, and mesh features.
```
Comment (pjcozzi) [inline CHANGES.md]: Is this still right?  I thought `mesh` features were removed.

File: Source/Shaders/Builtin/Functions/writeDepthClampedToFarPlane.glsl [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ * </p>
+ *
+ * @name czm_writeDepthClampedToFarPlane
+ * @glslFunction
+ *
+ * @example
+ * gl_FragColor = color;
+ * czm_writeDepthClampedToFarPlane();
+ *
+ * @see czm_writeDepthClampedToFarPlane
```
Comment (pjcozzi) [inline Source/Shaders/Builtin/Functions/writeDepthClampedToFarPlane.glsl]: This `@see` references the same function.

File: Source/Scene/ClassificationModel.js [hunk 10]
```diff
@@ -0,0 +1,1869 @@
...
+     * @param {Boolean} [options.debugWireframe=false] For debugging only. Draws the model in wireframe.
+     * @param {DistanceDisplayCondition} [options.distanceDisplayCondition] The condition specifying at what distance from the camera that this model will be displayed.
+     * @param {ClassificationType} [options.classificationType] What this model will classify.
+     *
+     * @exception {DeveloperError} bgltf is not a valid Binary glTF file.
+     * @exception {DeveloperError} Only glTF Binary version 1 is supported.
+     *
+     * @demo {@link http://cesiumjs.org/Cesium/Apps/Sandcastle/index.html?src=3D%20Models.html|Cesium Sandcastle Models Demo}
+     */
+    function ClassificationModel(options) {
```
Comment (pjcozzi) [inline Source/Scene/ClassificationModel.js]: This still needs tests, right?  We should test this as an isolated unit, and not just rely on the 3D Tiles tests.

File: Source/Scene/ClassificationModel.js [hunk 11]
```diff
@@ -0,0 +1,1869 @@
...
+        './getAttributeOrUniformBySemantic',
+        './HeightReference',
+        './ModelMaterial',
+        './ModelMesh',
+        './ModelNode',
+        './SceneMode',
+        './Vector3DTileBatch',
+        './Vector3DTilePrimitive'
+    ], function(
+        arraySlice,
```
Comment (pjcozzi) [inline Source/Scene/ClassificationModel.js]: Did you remove unused defines after the latest refactor?

File: Source/Scene/ClassificationModel.js [hunk 12]
```diff
@@ -0,0 +1,1869 @@
...
+    if (!FeatureDetection.supportsTypedArrays()) {
+        return {};
+    }
+
+    var boundingSphereCartesian3Scratch = new Cartesian3();
+
+    var ModelState = {
+        NEEDS_LOAD : 0,
+        LOADING : 1,
+        LOADED : 2,  // Renderable, but textures can still be pending when incrementallyLoadTextures is true.
```
Comment (pjcozzi) [inline Source/Scene/ClassificationModel.js]: Comment doesn't matter in this context, right?

File: Source/Scene/ClassificationModel.js [hunk 13]
```diff
@@ -0,0 +1,1869 @@
...
+     */
+    function ClassificationModel(options) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        var gltf = options.gltf;
+        if (gltf instanceof ArrayBuffer) {
+            gltf = new Uint8Array(gltf);
+        }
+
+        if (gltf instanceof Uint8Array) {
```
Comment (pjcozzi) [inline Source/Scene/ClassificationModel.js]: Should this only support Binary glTF?

File: Source/Scene/ClassificationModel.js [hunk 14]
```diff
@@ -0,0 +1,1869 @@
...
+         * </p>
+         *
+         * @type {Boolean}
+         *
+         * @default false
+         */
+        this.debugWireframe = defaultValue(options.debugWireframe, false);
+        this._debugWireframe = false;
+
+        this._distanceDisplayCondition = options.distanceDisplayCondition;
```
Comment (pjcozzi) [inline Source/Scene/ClassificationModel.js]: Remind me where this is used via 3D Tiles?

File: Source/Scene/ClassificationModel.js [hunk 15]
```diff
@@ -0,0 +1,1869 @@
...
+        this._maxDirtyNumber = 0;                  // Used in place of a dirty boolean flag to avoid an extra graph traversal
+
+        this._runtimeNode = {
+            matrix : undefined,
+            translation : undefined,
+            rotation : undefined,
+            scale : undefined,
+
+            // Per-node show inherited from parent
+            computedShow : true,
```
Comment (pjcozzi) [inline Source/Scene/ClassificationModel.js]: Here and below, does this still need to be simplified give that we will only support one mesh?  Did we confirm that only one primitive is OK as well?

File: Source/Scene/ClassificationModel.js [hunk 16]
```diff
@@ -0,0 +1,1869 @@
...
+                return this._geometryByteLength;
+            }
+        },
+
+        /**
+         * Gets the model's texture memory in bytes.
+         *
+         * @private
+         */
+        texturesByteLength : {
```
Comment (pjcozzi) [inline Source/Scene/ClassificationModel.js]: Just return zero?

File: Source/Scene/ClassificationModel.js [hunk 17]
```diff
@@ -0,0 +1,1869 @@
...
+        }
+
+        var gltfNode = nodes[0];
+        var runtimeNode = model._runtimeNode;
+        if (defined(gltfNode.matrix)) {
+            runtimeNode.matrix = Matrix4.fromColumnMajorArray(gltfNode.matrix);
+        } else {
+            // TRS converted to Cesium types
+            var rotation = gltfNode.rotation;
+            runtimeNode.translation = Cartesian3.from
```
Comment (pjcozzi) [inline Source/Scene/ClassificationModel.js]: Seems like once this is trimmed down for one node/primitive, there are some obvious things that can be factored into a common file such as the quantization functions.  I wonder if extension handling can also be made more simple here since only quantization and RTC need to be supported.

Also remind me, can we completely ignore the shaders since we are assuming no alpha texture / discard?

File: CHANGES.md (hunk 8 above)
Comment (bagnell) [reply CHANGES.md]: They were removed in the other PR.

File: Source/Scene/ClassificationModel.js (hunk 11 above)
Comment (bagnell) [reply Source/Scene/ClassificationModel.js]: Removed.

File: Source/Scene/ClassificationModel.js (hunk 10 above)
Comment (bagnell) [reply Source/Scene/ClassificationModel.js]: I started writing tests here but it required duplicating a ton of code from `Batched3DModel3DTileContent` in the spec.

**Comment by bagnell** [comment]
> Do you get test failures like this:

I don't see any test failure locally and the Travis tests pass. Also, running the tests from the link you posted give strange failures like:

Core/Cartesian3 renders with external gltf

**Comment by bagnell** [comment]
The requirement for floating point positions and uint16 batch ids were removed as well.Different formats are required to support quantization.

File: Source/Scene/ClassificationModel.js (hunk 14 above)
Comment (bagnell) [reply Source/Scene/ClassificationModel.js]: I thought it was used, but was not. Removed.

**Comment by bagnell** [comment]
@pjcozzi I think this is ready. The only thing missing would be only supporting one primitive per mesh, but I could do that in a separate PR.

**Comment by pjcozzi** [comment]
> The only thing missing would be only supporting one primitive per mesh, but I could do that in a separate PR.

Separate PR is OK, but let's do it right away to keep the code streamlined.

File: Source/Scene/ClassificationModel.js [hunk 18]
```diff
@@ -0,0 +1,1807 @@
...
+                accessorMax = quantizedAttributes.decodedMax;
+            }
+        }
+        return {
+            min : accessorMin,
+            max : accessorMax
+        };
+    }
+
+    function computeBoundingSphere(model) {
```
Comment (pjcozzi) [inline Source/Scene/ClassificationModel.js]: Are you going to move some of these functions to a common file?  Or do you want to do that in a separate PR?

File: Source/Scene/ClassificationModel.js (hunk 10 above)
Comment (pjcozzi) [reply Source/Scene/ClassificationModel.js]: Is it possible that when you create a common file for both `Model` implementations that it will because easier to write unit tests since you can also write tests against that?

File: Source/Scene/ClassificationModel.js (hunk 17 above)
Comment (bagnell) [reply Source/Scene/ClassificationModel.js]: RTC, binary glTF, and quantization are the extensions that are supported. The other glTF extensions are for the material and technique which we don't even parse.

We now also completely ignore the shaders.

**Comment by bagnell** [comment]
@pjcozzi This is ready for another look. I added the restriction that only one primitive per mesh is supported, factored out some of the common code with Model, and added ClassificationModel tests.

File: Source/Scene/ClassificationModel.js [hunk 19]
```diff
@@ -0,0 +1,1209 @@
...
+            get : function() {
+                return this._classificationType;
+            }
+        }
+    });
+
+    var aMinScratch = new Cartesian3();
+    var aMaxScratch = new Cartesian3();
+
+    function computeBoundingSphere(model) {
```
Comment (pjcozzi) [inline Source/Scene/ClassificationModel.js]: Can this be a common function?

File: Source/Scene/ClassificationModel.js [hunk 20]
```diff
@@ -0,0 +1,1209 @@
...
+                return (indicesCount / 3);
+            case PrimitiveType.TRIANGLE_STRIP:
+            case PrimitiveType.TRIANGLE_FAN:
+                return Math.max(indicesCount - 2, 0);
+            default:
+                return 0;
+        }
+    }
+
+    function createCommand(model) {
```
Comment (pjcozzi) [inline Source/Scene/ClassificationModel.js]: Maybe rename this since it no longer creates a command...directly.

**Comment by pjcozzi** [comment]
Tests and Sandcastle example look good.  Just those trivial questions.

File: Source/Scene/ClassificationModel.js (hunk 19 above)
Comment (bagnell) [reply Source/Scene/ClassificationModel.js]: This function is different enough from model that I didn't create a common function. The only part I could pull out is finding the min and max.

**Comment by bagnell** [comment]
@pjcozzi This should be ready.

**Comment by pjcozzi** [comment]
Thanks!
